src/vm/contexts.rs

```diff
@@ -39,9 +43,184 @@ pub struct LocalContext <'a> {
+        Environment::new(self.global_context,
+                         self.contract_context,
+                         self.call_stack,
+                         Some(sender))
+    }
+
+    pub fn eval_read_only(&mut self, contract_name: &str, program: &str) -> Result<Value> {
+        let parsed = parser::parse(program)?;
+        if parsed.len() < 1 {
+            return Err(Error::new(ErrType::ParseError("Expected a program of atleast length 1".to_string())))
+        }
+
+        let contract = self.global_context.database.get_contract(contract_name)?;
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, &contract.contract_context, self.call_stack, self.sender.clone());
+            let local_context = LocalContext::new();
+            eval(&parsed[0], &mut nested_env, &local_context)
+        };
+        nested_context.database.roll_back();
+
+        result
+    }
+
+    pub fn execute_contract(&mut self, contract_name: &str, 
+                            tx_name: &str, args: &[SymbolicExpression]) -> Result<Value> {
+        let contract = self.global_context.database.get_contract(contract_name)?;
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, &contract.contract_context, self.call_stack, self.sender.clone());
+            contract.execute_transaction(tx_name, args, &mut nested_env)
+        };
+
+        nested_context.handle_tx_result(result)
+    }
+
+    pub fn execute_function_as_transaction(&mut self, function: &DefinedFunction, args: &[Value]) -> Result<Value> {
+        let function = match function {
+            DefinedFunction::Private(_) => Err(Error::new(ErrType::NonPublicFunction(format!("{}", function.get_identifier())))),
+            DefinedFunction::Public(f) => Ok(f)
+        }?;
+
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+
+        let result = {
+            let mut nested_env = Environment::new(&mut nested_context, self.contract_context, self.call_stack, self.sender.clone());
+
+            function.apply(args, &mut nested_env)
+        };
+
+        nested_context.handle_tx_result(result)
+    }
+
+    pub fn initialize_contract(&mut self, contract_name: &str, contract_content: &str) -> Result<()> {
+        let mut nested_context = GlobalContext::begin_from(&mut self.global_context.database);
+        let result = Contract::initialize(contract_name, contract_content,
+                                          &mut nested_context);
+        match result {
+            Ok(contract) => {
+                nested_context.database.insert_contract(contract_name, contract)?;
+                nested_context.commit();
+                Ok(())
+            },
+            Err(e) => {
```

Comment:
What are the possible failures here?  Can the database I/O fail due to transient issues (like lock contention)? If so, these need to be handled explicitly.  Generally speaking, it's probably best to treat most database I/O failures as panics unless we explicitly check that they can be recovered from (such as, in the case of lock contention, by re-trying with exponential backoff).

Reply:
`Contract::initialize` is evaluating the `define` statements, and any additional code in the contract initialization. Any interpreter error is possible in that `Err(e)` branch.

For `database.insert_contract` -- I think we should change that to a panic.